scrapy_poet/page_input_providers.py

```diff
@@ -54,49 +55,39 @@ def __init__(self, response: Response):
         """
 
     @abc.abstractmethod
-    def __call__(self):
+    def __call__(self, provided_classes: typing.Set[typing.Type]) -> typing.Dict[typing.Type, typing.Any]:
         """This method is responsible for building Page Input dependencies."""
 
 
-def register(provider_class: typing.Type[PageObjectInputProvider],
-             provided_class: typing.Type):
-    """This method registers a Page Object Input Provider in the providers
-    registry. It could be replaced by the use of the ``provides`` decorator.
-
-    Example:
-
-        register(ResponseDataProvider, ResponseData)
-    """
-    providers[provided_class] = provider_class
-
-
-def provides(provided_class: typing.Type):
+def provides():
     """This decorator should be used with classes that inherits from
     ``PageObjectInputProvider`` in order to automatically register them as
     providers.
 
     See ``ResponseDataProvider``'s implementation for an example.
     """
     def decorator(provider_class: typing.Type[PageObjectInputProvider]):
-        register(provider_class, provided_class)
+        providers.add(provider_class)
         return provider_class
 
     return decorator
 
 
-@provides(ResponseData)
+@provides()
 class ResponseDataProvider(PageObjectInputProvider):
     """This class provides ``web_poet.page_inputs.ResponseData`` instances."""
 
+    provided_classes = {ResponseData}
+
     def __init__(self, response: Response):
         """This class receives a Scrapy ``Response`` as a dependency."""
         self.response = response
 
-    def __call__(self):
-        """This method builds a ``ResponseData`` instance using a Scrapy
-        ``Response``.
-        """
-        return ResponseData(
-            url=self.response.url,
-            html=self.response.text
-        )
+    def __call__(self, *args, **kwargs):
```

Comment:
It could be better to receive provided_classes explicitly, what do you think?

Reply:
Absolutely, there's a lot of room for improvement on this pr as I was just pushing things from my local machine 😂

Reply:
@kmike, before committing this change, a couple of questions:

- [how] should I type annotate the `provided_classes` argument?
- should I assert `provided_classes == self.provided_classes` or `provided_classes == {ResponseData}`

Thinking here if it makes sense to have a class attribute named `provided_class` for single-object providers. Our injection middleware could check if wither attribute is defined and raise an error if more than one is defined. This could simplify things a little bit (or make them more complex, never know). What do you think?

Reply:
>  Indeed I would probably change this name to something more specific like to_provide_remaining, which is what the set represents: what classes has not been yet provided.

I don't like naming it as `to_provide_remaining` because it gives the idea that the provider might be called multiple times and it's not what's intended. Actually, calling the provider multiple times could be the source of issues since we "think" we're calling it with every dependency but suddenly, we're not.

> If there are no classes in the to_provide_remaining that the provider can provide, then it should do nothing. Rerturn the empty dict. This logic is not present in this provider.

I like this proposal. I'll try to implement it. 

> The reason for that is that there can be redundant providers. More than one provider could be providing the same type (e.g. Html data). Once one provider has provided this type, the rest should not try to provide it.

I'm not sure it should be possible to register multiple providers for the same type because that could make things a little bit more complicated or cause confusion: which one is going to be executed?

Reply:
> If there are no classes in the to_provide_remaining that the provider can provide, then it should do nothing. Rerturn the empty dict. This logic is not present in this provider.

I was going to implement this suggestion, but in the end I don't think it's necessary because the injection middleware (and its utils) take care of this use case.

`build_instances` method will only call a provider instance if a dependency is listed on its provided classes attribute, otherwise, it will just skip this provider. So no problem here and no need to handle this case in the provider itself because it won't ever be called.

Reply:
> build_instances method will only call a provider instance if a dependency is listed on its provided classes attribute, otherwise, it will just skip this provider. So no problem here and no need to handle this case in the provider itself because it won't ever be called.

An example where this won't be true:

```py 
class ProductProvider(PageObjectInputProvider):
  provided_classes = [Product, Html]

  def _call_(...):
    ...

class ArticleProvider(PageObjectInputProvider):
  provided_classes = [Article, Html]

  def _call_(...):
    ...
```

And then you have a callback in a spider:
```py
class MySpider(Spider):
  def parse(response, product: Product, article: Article, html: Html):
    pass
```

In such a case, `build_instances` will use both providers, and both providers will ask for Html. This is not right, only the first provider (whichever it is) should ask for the Html in this case. And for that you need the `to_provide_remaining` mechanism. 

There are other cases: overriding a provider. Imagine that you have a single provider that provides all the AutoExtract page objects:
```py
class AutoExtractProvider(PageObjectInputProvider):
  provided_classes = [Article, Product, JobPosting, Html]

  ....
```
But you want to override the provider for Product, because you are obtaining the product by other means. The overriding can be obtained just but placing your new provider before the original provider. But you need the `to_provide_remaining` mechanism to avoid both providers performing the request. 

IMO this mechanism is important thus. 

> I'm not sure it should be possible to register multiple providers for the same type because that could make things a little bit more complicated or cause confusion: which one is going to be executed?

I don't think is complicated. Simply the order will be respected. Is the same than with Scrapy extensions. Maybe we could use numbers like in Scrapy to make simpler the ordering. This can be done in a different PR, but in this one it think at least the ordering should be respected. 

Also note that not having order is even worse in terms of confusion, because in this case you'll never know which one will be executed. So results won't be necessary deterministic.

Reply:
I finished looking at the code and I see that `__call__` receives only the list of classes that the provider must take care of. Makes sense. Implementing then the proposal that allows overriding is just a matter of removing from this list those for which you already have an instance. Just modifying one line in the code. And `to_provide` is then a much better name :-)

Reply:
- renamed `provided_classes` argument to `to_provided` 
- respecting order of providers

Reply:
Addressed. Resolving.